integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,350 @@
+
+def dump_stats_summary(df, fname):
+    """
+    Write mean runtime and energy and the standard deviation of 
+    runtime and energy for each power limit in CSV format to the
+    file fname."""
+    means = df \
+            .groupby(level=0) \
+            .mean() \
+            .rename(lambda x: x+"_mean", axis=1)
+    stds = df \
+            .groupby(level=0) \
+            .mean() \
+            .rename(lambda x: x+"_stddev", axis=1)
+    pandas.concat([means, stds], axis=1).to_csv(fname)
+
+
+def policy_min_energy(
+        plrange,
+        enmodel,
+        rtmodel = None,
+        pltdp = None,
+        max_degradation = None):
+    """
+    Find the power limit over the range plrange (list-like) that
+    has the minimum predicted energy usage (according to the energy model
+    enmodel, an instance of PowerLimitModel), subject to the constraint
+    that its runtime does not exceed the runtime at power limit pltdp
+    by more than a factor of (1 + max_degradation), if max_degradation is
+    specified. Returns a dictionary with keys power, runtime, and energy,
+    and values the optimal power limit and the predicted runtime and energy
+    at that limit, respectively."""
+
+    if pltdp is None:
+        pltdp = max(plrange)
+
+    en_predictions = enmodel.batch_test(plrange)
+    if max_degradation is None:
+        # we don't need the runtime model in this case
+        best_energy, best_pl = min(zip(en_predictions, plrange))
+        if rtmodel:
+            best_runtime = rtmodel.test(best_pl)
+        else:
+            best_runtime = None
+        return {'power': best_pl,
+                'runtime': best_runtime,
+                'energy': best_energy}
+    else:
+        rt_predictions = rtmodel.batch_test(plrange)
+        rt_at_tdp = rtmodel.test(pltdp)
+        constrained_values = [
+                (energy, pl, runtime)
+                for pl, runtime, energy
+                in zip(plrange, rt_predictions, en_predictions)
+                if runtime <= (1 + max_degradation) * rt_at_tdp]
+        best_energy, best_pl, best_runtime = min(constrained_values)
+        return {'power': best_pl,
+                'runtime': best_runtime,
+                'energy': best_energy}
+
+
+def main(full_df,
+        region_filter,
+        dump_prefix,
+        min_pl,
+        max_pl,
+        tdp,
+        min_energy,
+        max_degradation):
+    """
+    The main function. full_df is a report collection dataframe,
+    region_filter is a list of regions to include, dump_prefix
+    a filename prefix for debugging output (if specified), min_pl,
+    max_pl, tdp are the minimum, maximum, and reference power limits,
+    respectively, min_energy is a flag indicating whether or not
+    minimum energy is sought not subject to a constraint on runtime
+    (True if runtime should be ignored) and max_degradation specifies
+    what maximum runtime degradation is accepted."""
+    df = extract_columns(full_df, args.region_filter)
```

Comment:
how is args defined? I think this should just be region_filter

Reply:
Good catch.